src/passes/Precompute.cpp

```diff
@@ -743,131 +745,169 @@ struct Precompute
-        // returned value's type. To handle that, if we precomputed a value and
-        // if it has the wrong type then precompute it again without looking
-        // through to the fallthrough.
-        if (values.isConcrete() &&
-            !Type::isSubType(values.getType(), set->value->type)) {
-          values = precomputeValue(set->value);
-        }
+
+    // The work list, which will contain sets and gets that have just been
+    // found to have a constant value. As we only add them to the list when they
+    // are found to be constant, each can only be added once, and a simple
+    // vector is enough here (which we can make a small vector to avoid any
+    // allocation in small-enough functions).
+    SmallVector<Expression*, 10> work;
+
+    // Given a set, see if it has a constant value. If so, note that on
+    // setValues and add to the work list.
+    auto checkConstantSet = [&](LocalSet* set) {
+      if (setValues.count(set)) {
+        // Already known to be constant.
+        return;
+      }
+
+      // Precompute the value. Note that this executes the code from scratch
+      // each time we reach this point, and so we need to be careful about
+      // repeating side effects if those side effects are expressed *in the
+      // value*. A case where that can happen is GC data (each struct.new
+      // creates a new, unique struct, even if the data is equal), and so
+      // PrecomputingExpressionRunner has special logic to make sure that
+      // reference identity is preserved properly.
+      //
+      // (Other side effects are fine; if an expression does a call and we
+      // somehow know the entire expression precomputes to a 42, then we can
+      // propagate that 42 along to the users, regardless of whatever the call
+      // did globally.)
+      auto values = precomputeValue(
+        Properties::getFallthrough(set->value, getPassOptions(), *getModule()));
+
+      // Fix up the value. The computation we just did was to look at the
+      // fallthrough, then precompute that; that looks through expressions
+      // that pass through the value. Normally that does not matter here,
+      // for example, (block .. (value)) returns the value unmodified.
+      // However, some things change the type, for example RefAsNonNull has
+      // a non-null type, while its input may be nullable. That does not
+      // matter either, as if we managed to precompute it then the value had
+      // the more specific (in this example, non-nullable) type. But there
+      // is a situation where this can cause an issue: RefCast. An attempt to
+      // perform a "bad" cast, say of a function to a struct, is a case where
+      // the fallthrough value's type is very different than the actually
+      // returned value's type. To handle that, if we precomputed a value and
+      // if it has the wrong type then precompute it again without looking
+      // through to the fallthrough.
```

Comment:
In this case the cast will fail, so what useful work can this second precompute do?

Reply:
It will guarantee that we get something of the proper type that we can flow forward. Though if the type was wrong then I think it will always fail, is that what you are saying?

That seems to make sense, but as this code is unchanged in this PR, I'd like to consider it separately.

Reply:
Right, I was thinking that this would always fail. Considering it separately sounds good to me.